Restructure Circle CI config.yml to use Workflows

pavlospt opened the pull request:
This PR leverages CircleCI Workflows in order to build samples & run tests in parallel.

It first builds the whole project and then starts to run the following jobs in parallel:

1) `sample` build with BUCK
2) `sample-barebones` build with BUCK
3) `sample-codelab` build with BUCK
4) `sample-kotlin` build with Gradle
5) `litho-it` tests run with BUCK
6) `litho-it-powermock` tests run with BUCK
7) Tests run with Gradle

Every job that runs tests, also stores test results and artifacts. Note on this, the previous `regex` used was not grepping tests results from `litho-it` & `litho-it-powermock`. In order to save those results as well, I modified the jobs that run tests using `BUCK` to output the results in a `*.xml` file respectively.

## Structure of `config.yml` file

* Declare any references to be used in following steps
* Declare Jobs that will run in Workflow
  -  Each Job consists of:
     - Container Configuration
     - Workspace attaching
     - Cache restoration
     - Main work execution (checkout code/build/test runs)
     - Store tests results (optional)
     - Store tests artifacts
* Declare the Workflow alongside Jobs execution order and Jobs inter-dependencies

## Points to improve

* Currently `Workflow` is based on @passy 's Docker image specially created for `Litho`: 
1) Based on `circle-ci-android` Docker image
2) Installs `BUCK`
3) Installs `androidSdk`
4) Warms up some caches in order to reduce build time on the first job.

Although this is great, Circle CI caches images on `containers` so it was pretty hard to end up on a container with that specific Docker image cached. It did happen though a few times (luckily enough there were some times I needed quick feedback loop and the container had the image cached), effectively letting the Job setup the environment in 1 second. 

By switching to `circle-ci-android` Docker image and pulling inside a configuration job, that would of course run as a first Job in the `Workflow` would drammatically increase our chances to find a cached Docker image and decrease the 5-6 mins of environment setup to 1 second. Considering the number of jobs in the `Workflow` (10 atm) it can help reduce the average completion time at an order of magnitude 😃 

## Interesting facts about this PR

* It took me around 3 days or so 🕵️
* Ultimately final working version in my branch on my fork consisted of 59 fixup commits 😂 
* It took 269 job runs to finally get it to work 😮
* Average `Workflow` run is 40 mins (will explain below some tips that can improve the time) ⏲
* CI provisioning is a tough task 😂 

## Some related pics

### Cached image inside container
![Cached image inside container](https://user-images.githubusercontent.com/1070871/39297768-10c17ad8-494d-11e8-8884-a420b9fb8e72.png)

### Final success Workflow run
![Final success Workflow run](https://user-images.githubusercontent.com/1070871/39297781-16b11a8e-494d-11e8-9f93-cc8ae944e654.png)

passy commented:
AMAZING!

Thank you for working on this. The point about image is well made. I hope that Circle will give us better control about image propagation at some point because right now it is a gamble for every instance whether or not an image will be in cache or require downloading.

Moving to a standard image, however, would not help us. We were there before (hi, Circle 1.0) and it was awful. This is mainly due to us having to build buck every time and downloading all the Android dependencies which not only takes a lot of time in perfect conditions, but also has many external dependencies which can (and will) fail.

I'll do some tests on a branch here first to see that everything works as expected. I'd also like to check at least briefly if there are some ways to speed this up - perhaps by keeping our image closer to the circle base image. It's a bit annoying that this would actually increase the build time while it clearly should do the exact opposite given that everything is in parallel now.

passy commented:
One idea: Does `attach_workspace` work across different containers? Perhaps we could get away with using our big base image only for some of the tasks and use the standard `circle-ci-android` for others.

pavlospt commented:
@passy glad you liked it! I am also interested to find out if there are things that could be improved!! 

In case we can get the built version of BUCK copied inside the `workspace` and reference it from there, as well as the `androidSdk` it should work! My thoughts after that step were to also move the BUCK download task inside a Job that would persist it! (did not know that it had so many deps! :P )

passy commented:
Copying the two things over sounds like a great solution to it. At that
point we wouldn't need anything from the image and could just use the
standard one.

On Thu, 26 Apr 2018, 12:57 Pavlos-Petros Tournaris, <
notifications@github.com> wrote:

> @passy <https://github.com/passy> glad you liked it! I am also interested
> to find out if there are things that could be improved!!
>
> In case we can get the built version of BUCK copied inside the workspace
> and reference it from there, as well as the androidSdk it should work! My
> thoughts after that step were to also move the BUCK download task inside a
> Job that would persist it! (did not know that it had so many deps! :P )
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/facebook/litho/pull/350#issuecomment-384614539>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAmshF85PlPVLIZnrLNyrqQEqGOp_sSks5tsbY0gaJpZM4Tk09j>
> .
>
-- 
timezone: eu/london | utc+1
@passy

pavlospt commented:
@passy that's the first time I see it. This should only happen in case it failed to execute `checkout_code` which in turn should have stopped the workflow.

pavlospt commented:
@passy pushed a new commit dealing with the points we have discussed above: 

Specifically:
* I created another container configuration to use CircleCI's android image, which proved our hypothesis to be correct. Image is cached in every container, haven't had a single image cache miss.
* After the first job is run, I am moving BUCK, AndroidSDK and AndroidNDK inside workspace, in order to be able to use CircleCI's android image.
* I updated the cache key in order to perform a cache bust, due to the fact that the previous configuration had already caches BUCK and android-sdk folders, which was not needed. (Unfortunately in Circle 2.0, cache busting only works by updating the cache key, that's why I decided to proceed by adding a version on it, which can help us easily update it whenever we need to run without cache. Another possible solution would be to include another file holding that version and use `chksum` on it to have the same effect)
* New container configuration uses environment vars configured from the `workspace` folder.

Observations made: 
* Noticed that the initial Job will always take up a lot time, since it spins up the custom Docker image and also has to perform a move of those folders inside workspace.
* Persist to workspace will take some time for the `checkout_code` job due to the size of `buck` / `android-sdk` and `android-ndk` folders.
* Cache restoration time, has been drammatically reduced, alongside its size. I expect no OOM errors for that specific Job now. (They appeared sporadically previously)
* `Build` and `buck`-powered Jobs take only 4 mins to run, which is pretty amazing considering they need to attach to a 4.4 GB workspace, restore cache, run tests, store test results and finally store artifacts.
* Gradle tests run is still taking up a lot of time (Not sure what we can do for that case, but I think it worths checking out. Especially if no other module other than `litho-it` & `litho-it-powermock` contains tests, might worth running only them with Gradle as well) 

### Workflow execution after update
![screen shot 2018-04-27 at 10 23 16](https://user-images.githubusercontent.com/1070871/39350145-f00876d4-4a05-11e8-807c-f3845ab9f948.png)

### Build job
![screen shot 2018-04-27 at 10 23 41](https://user-images.githubusercontent.com/1070871/39350406-c3909f4a-4a06-11e8-9f5d-b5cc9c61c4fe.png)

### BUCK-powered sample build
![screen shot 2018-04-27 at 10 24 04](https://user-images.githubusercontent.com/1070871/39350422-ceaaf8bc-4a06-11e8-88ca-9b795e8ec898.png)

### BUCK-powered tests run
![screen shot 2018-04-27 at 10 24 13](https://user-images.githubusercontent.com/1070871/39350441-e1465fa2-4a06-11e8-9bcf-636d4148718f.png)

### Gradle-powered tests run
![screen shot 2018-04-27 at 10 23 49](https://user-images.githubusercontent.com/1070871/39350462-ee5b24de-4a06-11e8-990c-1b7792993746.png)

passy commented:
This is great work! Thanks again for working on this. I'd really love to ship this straight away, but I'm worried about effectively doubling our build times.

The main issue is clearly that the workspace persistence is taking wayyy too long. I'm not sure if there's much we can do. I haven't checked how much of the space is taken up by the buck installation - I'm sure we could trim this down to the actually needed files. Perhaps there's a build instruction somewhere to have a JAR + Wrapper script or something that we could get away with.

The biggest part is almost certainly the NDK + SDK. There's a lot of stuff we could probably avoid downloading and copying over if we could reuse the bits that come with the Circle images. That would essentially mean looking into https://github.com/facebook/litho/blob/master/scripts/circle-ci-android-setup.sh and see if we can remove downloads - or even better - use whatever Circle has installed for us instead.

Do you think that might be worth looking into to shave off some minutes from the workspace times?

passy commented:
![](https://media1.tenor.com/images/07feba4572471b21fd4258b6af83c9c4/tenor.gif?itemid=4158682)

pavlospt commented:
@passy Summing up on the actions followed for the speed performance optimization: 

1. Removed custom CircleCI Docker Image 
2. Using internal Android SDK after downloading what's needed for the project to run.
3. Downloading Android NDK on the first Job run
4. Downloading and building BUCK, also on the first Job run
5. Caching NDK, SDK and BUCK (BUCK is fetched from master)
6. Added `scripts/apt-get-android-deps.sh` in order to download build dependencies for BUCK

Also, rewrote `config.yml` in a `facebook/react-native` CircleCI configuration approach! 

Generally, as you have seen it is a lot faster than previously!

passy commented:
This is so cool. It'll take me awhile to make this work again internally, where our CI instances have way older deps installed by default. I'll likely have to revert some of the changes to the setup script and make sure we only skip the install steps if they're already provided by the infra.

pavlospt commented:
@passy Do your CI instances use the same script? Take your time! Let me know If i can help anywhere!

passy commented:
Yes, they currently do.
